libbeat/autodiscover/providers/docker/docker.go

```diff
@@ -0,0 +1,136 @@
+package docker
+
+import (
+	"github.com/elastic/beats/libbeat/autodiscover"
+	"github.com/elastic/beats/libbeat/autodiscover/template"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/common/docker"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+func init() {
+	autodiscover.Registry.AddProvider("docker", DockerAutodiscoveryBuilder)
+}
+
+type DockerAutodiscovery struct {
+	config        *Config
+	bus           bus.Bus
+	watcher       docker.Watcher
+	templates     *template.Mapper
+	stop          chan interface{}
+	startListener bus.Listener
+	stopListener  bus.Listener
+}
+
+func DockerAutodiscoveryBuilder(bus bus.Bus, c *common.Config) (autodiscover.Provider, error) {
+	config := defaultConfig()
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	mapper, err := template.NewConfigMapper(config.Templates)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher, err := docker.NewWatcher(config.Host, config.TLS)
+	if err != nil {
+		return nil, err
+	}
+
+	start := watcher.ListenStart()
+	stop := watcher.ListenStop()
+
+	if err := watcher.Start(); err != nil {
+		return nil, err
+	}
+
+	return &DockerAutodiscovery{
+		config:        config,
+		bus:           bus,
+		templates:     mapper,
+		watcher:       watcher,
+		stop:          make(chan interface{}),
+		startListener: start,
+		stopListener:  stop,
+	}, nil
+}
+
+func (d *DockerAutodiscovery) Start() {
+	go func() {
+		for {
+			select {
+			case <-d.stop:
+				d.startListener.Stop()
+				d.stopListener.Stop()
+				return
+
+			case event := <-d.startListener.Events():
+				d.emitContainer(event, "start")
+
+			case event := <-d.stopListener.Events():
+				d.emitContainer(event, "stop")
+			}
+		}
+	}()
+}
+
+func (d *DockerAutodiscovery) emitContainer(event bus.Event, flag string) {
+	container, ok := event["container"].(*docker.Container)
+	if !ok {
+		logp.Err("Couldn't get a container from watcher event")
+		return
+	}
+
+	meta := common.MapStr{
+		"container": common.MapStr{
+			"id":     container.ID,
+			"name":   container.Name,
+			"image":  container.Image,
+			"labels": container.Labels,
+		},
+	}
+
+	// Emit container info
+	d.emit(bus.Event{
+		flag:     true,
+		"host":   container.IPAddresses[0],
```

Comment:
can it happen that there is no IPAdress -> panic? Just double checking.

Reply:
good point, It should not happen, but will add some defensive code here